feat(arithmetic): Add field support to the arithmetic parser

wmak opened the pull request:
- There will only be a few fields that make sense to perform arithmetic
  on, so using an allowlist to prevent other fields from being used
- We'll want the list of fields in arithmetic to either validate that
  they're selected, or to make sure we can group by them if an aggregate
  is selected

wmak committed:
feat(arithmetic): Add field support to the arithmetic parser

- There will only be a few fields that make sense to perform arithmetic
  on, so using an allowlist to prevent other fields from being used
- We'll want the list of fields in arithmetic to either validate that
  they're selected, or to make sure we can group by them if an aggregate
  is selected

wmak commented on tests/sentry/discover/test_arithmetic.py:
```diff
@@ -56,35 +65,40 @@ def test_simple_arithmetic(a, op, b):
-    result.lhs.rhs = 34.0
-    result.rhs = 56.0
+    result, _ = parse_arithmetic("12 / 34 - 56")
+    assert result.operator == "minus"
+    assert isinstance(result.lhs, Operation)
+    assert result.lhs.operator == "divide"
+    assert result.lhs.lhs == 12.0
+    assert result.lhs.rhs == 34.0
+    assert result.rhs == 56.0
```
I'm very good at writing tests 🤡

k-fish replied:
lollll

Zylphrex replied:
The best test writer 💯

Zylphrex commented on src/sentry/discover/arithmetic.py:
```diff
@@ -65,12 +90,22 @@ class ArithmeticVisitor(NodeVisitor):
-    unwrapped_exceptions = (MaxOperatorError,)
+    unwrapped_exceptions = (ArithmeticError,)
+
+    allowlist = {
```
are we not allowing measurements in this pass?

wmak replied:
lol no i just forgot, thanks :+1:

Zylphrex commented on src/sentry/discover/arithmetic.py:
```diff
@@ -20,15 +22,37 @@ class ArithmeticParseError(ArithmeticError):
+    pass
+
+
+OperationSideType = Union["Operation", float]
```
I might be missing something but this `Union` should include field names right? Since `transaction.duration` is neither an `Operation` nor a `float`

Zylphrex reviewed (approved):
A few minor questions, but otherwise LGTM

wmak committed:
ref: Addressing PR comments

- Allowing measurements, and refactoring the allowlist to check that the
  field is a valid span/measurement or is just transaction.duration
- Fixing OperationSideType's type since it can be a str as well

wmak committed:
ref: Fixing tests + switching back to allowlist
